### PR TITLE
integrate unit id in catalogue category , item and catalogue items  #243

### LINKED ISSUE #243: Add unit IDs to catalogue categories, catalogue items, and items
Unit IDs should be added to catalogue categories, catalogue items, and items

### PR DESCRIPTION
**Pull request by joshuadkitenge**
## Description
see #243

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build

## Agile board tracking
closes #243

### CODE REVIEW

File: inventory_management_system_api/models/catalogue_item.py [hunk 1]
```diff
@@ -12,29 +12,31 @@
...
     name: str
     value: Any
+    unit_id: Optional[StringObjectIdField] = None
     unit: Optional[str] = None
 
     model_config = ConfigDict(populate_by_name=True)
 
 
-class CatalogueItemBase(BaseModel):
+class CatalogueItemBaseIn(BaseModel):
```
Comment (joelvdavies) [inline inventory_management_system_api/models/catalogue_item.py:39]:
```suggestion
class CatalogueItemIn(BaseModel):
```
Again I don't think we need the BaseIn and BaseOut, the properties should just be overridden in the In and Out models.

File: inventory_management_system_api/schemas/catalogue_category.py [hunk 2]
```diff
@@ -184,6 +185,9 @@ class CatalogueCategoryPatchRequestSchema(CatalogueCategoryPostRequestSchema):
         "elements but if it is not then it can only have catalogue categories as child elements.",
     )
     parent_id: Optional[str] = Field(default=None, description="The ID of the parent catalogue category")
+    catalogue_item_properties: Optional[List[CatalogueCategoryPostRequestPropertySchema]] = Field(
+        default=None, description="The properties that the catalogue items in this category could/should have"
+    )
```
Comment (joelvdavies) [inline inventory_management_system_api/schemas/catalogue_category.py:190]:
```suggestion
```
Already in CatalogueCategoryPostRequestSchema.

File: inventory_management_system_api/schemas/catalogue_item.py [hunk 3]
```diff
@@ -25,6 +25,7 @@ class PropertySchema(PropertyPostRequestSchema):
 
     name: str = Field(description="The name of the catalogue item property")
     unit: Optional[str] = Field(default=None, description="The unit of the property such as 'nm', 'mm', 'cm' etc")
+    unit_id: Optional[str] = Field(default=None, description="The ID of the unit")
```
Comment (joelvdavies) [inline inventory_management_system_api/schemas/catalogue_item.py:28]:
```suggestion
    unit_id: Optional[str] = Field(default=None, description="The ID of the unit")
    unit: Optional[str] = Field(default=None, description="The unit of the property such as 'nm', 'mm', 'cm' etc")
```
I would prefer unit_id to come first like the models.

**Review by joelvdavies — COMMENTED** [review]
A few comments on the database models/schemas.

File: inventory_management_system_api/services/catalogue_category_property.py [hunk 4]
```diff
@@ -89,7 +94,21 @@ def create(
...
+        if catalogue_item_property.unit_id is not None:
+            unit_id = catalogue_item_property.unit_id
+            unit = self._unit_repository.get(unit_id)
+            if not unit:
+                raise MissingRecordError(f"No unit found with ID: {unit_id}")
+        # pylint: enable=duplicate-code
+
+        catalogue_item_property_in = CatalogueItemPropertyIn(
+            **{**catalogue_item_property.model_dump(), "unit": unit.value}
+        )
```
Comment (joelvdavies) [inline inventory_management_system_api/services/catalogue_category_property.py:109]: This line causes a 500 error for the request
```
{
	"name": "test5",
	"type": "number",
	"mandatory": true,
	"allowed_values": null,
	"default_value": 0
}
```
as unit is None. We should have an e2e test to ensure this gets picked up.

**Review by joelvdavies — CHANGES_REQUESTED** [review]
Some more comments and a bug from going through services.

File: inventory_management_system_api/routers/v1/catalogue_category.py [hunk 5]
```diff
@@ -117,6 +117,14 @@ def create_catalogue_category(
         catalogue_category = catalogue_category_service.create(catalogue_category)
         return CatalogueCategorySchema(**catalogue_category.model_dump())
     except (MissingRecordError, InvalidObjectIdError) as exc:
+        if (
+            catalogue_category.catalogue_item_properties is not None
+            and any(str(property.unit_id) in str(exc) for property in catalogue_category.catalogue_item_properties)
+        ) or "units" in str(exc).lower():
+            message = "The specified unit does not exist"
+            logger.exception(message)
+            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=message) from exc
```
Comment (joelvdavies) [inline inventory_management_system_api/routers/v1/catalogue_category.py:126]:
```suggestion
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=message) from exc
```
We use this elsewhere if it can't process it due to the missing value, I believe 404's are more for the route parameters themselves rather than the body.

File: inventory_management_system_api/routers/v1/catalogue_category.py [hunk 6]
```diff
@@ -158,6 +166,13 @@ def partial_update_catalogue_category(
             message = "The specified parent catalogue category does not exist"
             logger.exception(message)
             raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=message) from exc
+        if (
+            catalogue_category.catalogue_item_properties is not None
+            and any(str(property.unit_id) in str(exc) for property in catalogue_category.catalogue_item_properties)
+        ) or "units" in str(exc).lower():
+            message = "The specified unit does not exist"
+            logger.exception(message)
+            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=message) from exc
```
Comment (joelvdavies) [inline inventory_management_system_api/routers/v1/catalogue_category.py:175]:
```suggestion
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=message) from exc
```

File: inventory_management_system_api/routers/v1/catalogue_category.py [hunk 7]
```diff
@@ -227,6 +242,14 @@ def create_catalogue_item_property(
...
         )
     except (MissingRecordError, InvalidObjectIdError) as exc:
+        if (
+            catalogue_item_property.unit_id is not None
+            and catalogue_item_property.unit_id in str(exc)
+            or "unit" in str(exc).lower()
+        ):
+            message = "The specified unit does not exist"
+            logger.exception(message)
+            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=message) from exc
```
Comment (joelvdavies) [inline inventory_management_system_api/routers/v1/catalogue_category.py:252]:
```suggestion
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=message) from exc
```

File: inventory_management_system_api/services/utils.py [hunk 8]
```diff
@@ -142,8 +142,9 @@ def _add_catalogue_item_property_units(
     :param supplied_properties: The supplied catalogue item properties.
     """
     logger.info("Adding the units to the supplied properties")
-    for supplied_property_id, supplied_property in supplied_properties.items():
-        supplied_property["unit"] = defined_properties[supplied_property_id]["unit"]
+    for supplied_property_name, supplied_property in supplied_properties.items():
+        supplied_property["unit"] = defined_properties[supplied_property_name]["unit"]
+        supplied_property["unit_id"] = defined_properties[supplied_property_name]["unit_id"]
```
Comment (joelvdavies) [inline inventory_management_system_api/services/utils.py:147]:
```suggestion
        supplied_property["unit_id"] = defined_properties[supplied_property_name]["unit_id"]
        supplied_property["unit"] = defined_properties[supplied_property_name]["unit"]
```
Just for consistency.

File: test/unit/services/test_catalogue_category.py
[the remark this answers is not in the document]
Comment (joshuadkitenge) [reply test/unit/services/test_catalogue_category.py:1]: e2e aswell

File: test/conftest.py [hunk 9]
```diff
@@ -30,29 +31,59 @@
...
-    `properties_with_ids` are provided.
+    `properties_with_ids` are provided. If units are provided, it matches and assigns unit IDs to properties
+    based on unit names.
 
     :param properties_with_ids: The list of properties with IDs. These are typically the catalogue category
     properties.
     :param properties_without_ids: The list of properties without IDs. These can be catalogue category, catalogue item
         or item properties.
+    :param units: The list of units used in the catalogue item properties
     :return: The list of properties with the added IDs.
```
Comment (joelvdavies) [inline test/conftest.py:48]:
```suggestion
    :param properties_with_ids: The list of properties with IDs. These are typically the catalogue category
                                                 properties.
    :param properties_without_ids: The list of properties without IDs. These can be catalogue category, catalogue item
                                                       or item properties.
    :param units: The list of units used in the catalogue item properties
    :return: The list of properties with the added IDs.
```

File: test/conftest.py [hunk 10]
```diff
@@ -30,29 +31,59 @@
...
-def add_ids_to_properties(properties_with_ids: Optional[list], properties_without_ids: list):
+def add_ids_to_properties(
+    properties_with_ids: Optional[list], properties_without_ids: list, units: Optional[list] = None
+):
     """
     A tests method for adding the IDs from the properties in `properties_with_ids` as the IDs to the properties in
     `properties_without_ids` based on matching names. Unique IDs are generated for each property if no
-    `properties_with_ids` are provided.
+    `properties_with_ids` are provided. If units are provided, it matches and assigns unit IDs to properties
+    based on unit names.
```
Comment (joelvdavies) [inline test/conftest.py:41]:
```suggestion
    `properties_with_ids` are provided. If units are provided, it matches and assigns unit IDs to properties
    based on unit values.
```

File: test/conftest.py [hunk 11]
```diff
@@ -30,29 +31,59 @@
...
+                            if property_without_id.get("unit") == unit["value"]:
+                                unit_id = unit["id"]
+                    else:
+                        unit_id = str(ObjectId())
+            # pylint: enable=too-many-nested-blocks
+            else:
+                unit_id = property_without_id["unit_id"]
 
-        properties.append({**property_without_id, "id": prop_id})
+        properties.append({**property_without_id, "id": prop_id, "unit_id": unit_id})
```
Comment (joelvdavies) [inline test/conftest.py:87]: I personally find
```
   # pylint: disable=too-many-nested-blocks
    for property_without_id in properties_without_ids:
        prop_id = None
        unit_id = None

        if properties_with_ids:
            # Match up property and unit ids
            for property_with_id in properties_with_ids:
                if property_with_id["name"] == property_without_id["name"]:
                    prop_id = property_with_id["id"]

                if property_with_id.get("unit") == property_without_id.get("unit"):
                    unit_id = property_with_id["unit_id"]
        else:
            # Generate a new property id and lookup the unit id from the units list
            prop_id = str(ObjectId())

            if property_without_id.get("unit") is not None:
                if property_without_id.get("unit_id") is None:
                    for unit in units:
                        if property_without_id["unit"] == unit["value"]:
                            unit_id = unit["id"]
                            break
                # pylint: enable=too-many-nested-blocks
                else:
                    unit_id = property_without_id["unit_id"]

        properties.append({**property_without_id, "id": prop_id, "unit_id": unit_id})

    return properties
```
easier to read

**Review by joelvdavies — COMMENTED** [review]
A few more comments before I look at the tests.

File: test/e2e/test_catalogue_category_property.py [hunk 12]
```diff
@@ -137,12 +139,29 @@ def setup(self, test_client):
...
+
+        self.units = [unit_mm]
+
+        response = self.test_client.post(
+            "/v1/catalogue-categories",
+            json={
+                **CATALOGUE_CATEGORY_POST_A,
+                "catalogue_item_properties": add_ids_to_properties(
+                    None, CATALOGUE_CATEGORY_POST_A["catalogue_item_properties"], self.units
+                ),
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_category_property.py:161]: I think we need a separate function just for units here actually, it works because the schema will ignore extra parameters but I think for e2e tests its best if we only supply what is needed, and here it will add ids and will still have the unit inside the request, whereas it should effectively replace the unit with unit id and leave the properties alone. I think the unit tests, or where its used in the output check is fine though as they do need both.

File: test/unit/services/test_catalogue_category.py [hunk 13]
```diff
@@ -307,6 +356,78 @@ def test_create_with_duplicate_property_names(catalogue_category_repository_mock
     )
 
 
+def test_create_catalogue_item_properties_with_non_existent_unit_id(
```
Comment (joelvdavies) [inline test/unit/services/test_catalogue_category.py:359]: Sorry just spotted, we do, do invalid ids for service tests in some cases, it would be like test_create_with_invalid_catalogue_category_id in items, so they can be done here.

File: test/unit/services/test_catalogue_category_property.py [hunk 14]
```diff
@@ -50,8 +53,9 @@ def test_create(
     property with a default value
     """
     catalogue_category_id = str(ObjectId())
+    unit = UnitOut(id=str(ObjectId()), **UNIT_A)
     catalogue_item_property = CatalogueItemPropertyPostRequestSchema(
-        name="Property A", type="number", unit="mm", mandatory=mandatory, default_value=default_value
+        name="Property A", type="number", unit_id=unit.id, mandatory=mandatory, default_value=default_value
```
Comment (joelvdavies) [inline test/unit/services/test_catalogue_category_property.py:58]: I think the tests below are missing this change.

File: inventory_management_system_api/services/catalogue_category_property.py [hunk 15]
```diff
@@ -89,7 +94,22 @@ def create(
...
+        # pylint: disable=duplicate-code
+        if catalogue_item_property.unit_id is not None:
+            unit_id = catalogue_item_property.unit_id
+            unit = self._unit_repository.get(unit_id)
+            if not unit:
+                raise MissingRecordError(f"No unit found with ID: {unit_id}")
+        # pylint: enable=duplicate-code
+
+        # Ensure that unit is not None before accessing its value
+        unit_value = unit.value if unit else None
```
Comment (VKTB) [inline inventory_management_system_api/services/catalogue_category_property.py:108]: We can make this shorter
```suggestion
        unit_value = None
        if catalogue_item_property.unit_id is not None:
            # Obtain the specified unit value if a unit ID is given
            unit = self._unit_repository.get(catalogue_item_property.unit_id)
            if not unit:
                raise MissingRecordError(f"No unit found with ID: {catalogue_item_property.unit_id}")
            unit_value = unit.value
```

File: inventory_management_system_api/services/catalogue_category.py [hunk 16]
```diff
@@ -143,6 +163,40 @@ def update(
...
+        catalogue_item_properties_with_units = []
+        for catalogue_item_property in catalogue_item_properties:
+            # pylint: disable=duplicate-code
+            if catalogue_item_property.unit_id is not None:
+                unit_id = catalogue_item_property.unit_id
+                unit = self._unit_repository.get(unit_id)
+                if not unit:
+                    raise MissingRecordError(f"No unit found with ID: {unit_id}")
+                # pylint: enable=duplicate-code
+                # Copy unit ID and value to catalogue item property
```
Comment (VKTB) [inline inventory_management_system_api/services/catalogue_category.py:196]:
```suggestion
            if catalogue_item_property.unit_id is not None:
                unit = self._unit_repository.get(catalogue_item_property.unit_id)
                if not unit:
                    raise MissingRecordError(f"No unit found with ID: {catalogue_item_property.unit_id}")
                # Copy unit value to catalogue item property
```

**Review by VKTB — CHANGES_REQUESTED** [review]
Some comments before I look at the tests.

**Comment by VKTB** [comment]
I will review the tests once the requested changes have been addressed as they are likely to affect the tests.

File: inventory_management_system_api/routers/v1/catalogue_category.py [hunk 17]
```diff
@@ -117,6 +117,14 @@ def create_catalogue_category(
         catalogue_category = catalogue_category_service.create(catalogue_category)
         return CatalogueCategorySchema(**catalogue_category.model_dump())
     except (MissingRecordError, InvalidObjectIdError) as exc:
+        if (
+            catalogue_category.catalogue_item_properties is not None
+            and any(str(property.unit_id) in str(exc) for property in catalogue_category.catalogue_item_properties)
+        ) or "units" in str(exc).lower():
```
[the remark this answers is not in the document]
Comment (joshuadkitenge) [reply inventory_management_system_api/routers/v1/catalogue_category.py:123]: Good spot it should be "unit"

File: inventory_management_system_api/services/catalogue_category_property.py [hunk 18]
```diff
@@ -89,7 +94,18 @@ def create(
             stored_catalogue_category.catalogue_item_properties + [catalogue_item_property]
         )
 
-        catalogue_item_property_in = CatalogueItemPropertyIn(**catalogue_item_property.model_dump())
+        # Obtain the specified unit if an id is given
+        unit_value = None
+        if catalogue_item_property.unit_id is not None:
+            # Obtain the specified unit value if a unit ID is given
```
Comment (joelvdavies) [inline inventory_management_system_api/services/catalogue_category_property.py:100]: I think only one of these comments are needed.

File: test/unit/services/test_catalogue_category.py [hunk 19]
```diff
@@ -307,6 +356,150 @@ def test_create_with_duplicate_property_names(catalogue_category_repository_mock
...
+            parent_id=None,
+            catalogue_item_properties=[],
+            created_time=MODEL_MIXINS_FIXED_DATETIME_NOW,
+            modified_time=MODEL_MIXINS_FIXED_DATETIME_NOW,
+        ),
+    )
+    # pylint: enable=duplicate-code
+
+    # Mock `get` to return the unit
+    test_helpers.mock_get(unit_repository_mock, None)
```
Comment (joelvdavies) [inline test/unit/services/test_catalogue_category.py:485]: This is just doing the same as the test above rather than checking for an InvalidObjectIdError. I realise now the test I saw is on its own and is for database integrity, so maybe this test doesn't make sense after all. I think it should just be removed again its tested elsewhere anyway, sorry 😅

File: test/e2e/conftest.py [hunk 20]
```diff
@@ -35,3 +36,45 @@ def fixture_cleanup_database_collections():
     database.systems.delete_many({})
     database.units.delete_many({})
     database.usage_statuses.delete_many({})
+
+
+def replace_unit_values_with_ids_in_properties(properties_without_ids: list, units: Optional[list]):
```
Comment (joelvdavies) [inline test/e2e/conftest.py:41]:
```suggestion
def replace_unit_values_with_ids_in_properties(properties_without_ids: list[dict], units: Optional[list]) -> list[dict]:
```

File: test/e2e/test_item.py [hunk 21]
```diff
@@ -82,12 +84,55 @@
...
+# pylint: enable=duplicate-code
+
+
 def test_create_item(test_client):
     """
     Test creating an item.
     """
-    response = test_client.post("/v1/catalogue-categories", json=CATALOGUE_CATEGORY_POST_A)
-    catalogue_category = response.json()
+    # pylint: disable=duplicate-code
```
Comment (joelvdavies) [inline test/e2e/test_item.py:133]:
```suggestion
    # pylint: disable=duplicate-code
```
Is this still needed, if so the one below should probably be removed.

File: test/e2e/test_catalogue_item.py [hunk 22]
```diff
@@ -1147,17 +1177,17 @@ def test_partial_update_catalogue_item_change_catalogue_category_id(test_client)
     Test moving a catalogue item to another catalogue category with the same properties without
     specifying any new properties.
     """
-    response = test_client.post("/v1/catalogue-categories", json=CATALOGUE_CATEGORY_POST_A)
-    catalogue_category_a = response.json()
-    response = test_client.post(
-        "/v1/catalogue-categories",
-        json={
+    # pylint: disable=duplicate-code
+    # units
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_item.py:1181]:
```suggestion
```

File: test/e2e/test_catalogue_category.py [hunk 23]
```diff
@@ -107,10 +109,33 @@ def _post_nested_catalogue_categories(test_client, entities: list[dict]):
 def _post_catalogue_categories(test_client):
     """Utility function for posting all mock systems defined at the top of this file"""
 
+    # units
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_category.py:112]:
```suggestion
```

File: test/e2e/test_catalogue_category.py [hunk 24]
```diff
@@ -1376,6 +1721,92 @@ def test_partial_update_catalogue_category_change_catalogue_item_properties_has_
     assert response.json()["detail"] == "Catalogue category has child elements and cannot be updated"
 
 
+def test_partial_update_catalogue_category_invalid_unit_id(test_client):
+    """
+    Test modifying a catalogue item property when there is an invalid unit ID.
+    """
+
+    # units
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_category.py:1729]:
```suggestion
```

File: test/e2e/test_catalogue_category.py [hunk 25]
```diff
@@ -1404,12 +1835,23 @@ def test_partial_update_catalogue_items_to_have_duplicate_property_names(test_cl
     """
     Test updating a catalogue category to have duplicate catalogue item property names
     """
-    response = test_client.post("/v1/catalogue-categories", json=CATALOGUE_CATEGORY_POST_C)
+    # units
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_category.py:1838]:
```suggestion
```

File: test/e2e/test_catalogue_category.py [hunk 26]
```diff
@@ -1376,6 +1721,92 @@ def test_partial_update_catalogue_category_change_catalogue_item_properties_has_
...
+    assert response.status_code == 422
+    assert response.json()["detail"] == "The specified unit does not exist"
+
+
+def test_partial_update_catalogue_category_non_existent_unit_id(test_client):
+    """
+    Test modifying a catalogue item property when there is an non existent unit ID.
+    """
+
+    # units
```
Comment (joelvdavies) [inline test/e2e/test_catalogue_category.py:1772]:
```suggestion
```

File: test/e2e/test_catalogue_category.py (hunk 23 above)
Comment (joshuadkitenge) [reply test/e2e/test_catalogue_category.py:112]: why do I need to get rid of this ??

File: inventory_management_system_api/services/catalogue_category.py [hunk 27]
```diff
@@ -56,12 +64,24 @@ def create(self, catalogue_category: CatalogueCategoryPostRequestSchema) -> Cata
         if parent_catalogue_category and parent_catalogue_category.is_leaf:
             raise LeafCategoryError("Cannot add catalogue category to a leaf parent catalogue category")
 
+        catalogue_item_properties = []
         if catalogue_category.catalogue_item_properties:
             utils.check_duplicate_catalogue_item_property_names(catalogue_category.catalogue_item_properties)
 
+            catalogue_item_properties = self._add_catalogue_item_property_units(
+                catalogue_category.catalogue_item_properties
```
[the remark this answers is not in the document]
Comment (VKTB) [reply inventory_management_system_api/services/catalogue_category.py:76]: This is still complaining

[the remark this answers is not in the document]
Comment (VKTB) [reply inventory_management_system_api/services/catalogue_category.py:76]: @joshuadkitenge This is still complaining

[the remark this answers is not in the document]
Comment (VKTB) [reply inventory_management_system_api/services/catalogue_category.py:76]: `Expected type 'list[CatalogueItemPropertyPostRequestSchema | CatalogueItemPropertyOut]', got 'list[CatalogueCategoryPostRequestPropertySchema]' instead `

[the remark this answers is not in the document]
Comment (joelvdavies) [reply inventory_management_system_api/services/catalogue_category.py:76]: There is a vscode typechecker in the settings
image

Although it does also report on some issues I hadn't noticed before that presumably pycharm is also missing.

File: test/e2e/test_catalogue_category.py [hunk 28]
```diff
@@ -129,6 +152,19 @@ def _post_n_catalogue_categories(test_client, number):
     )
 
 
+def _post_units(test_client):
```
Comment (VKTB) [inline test/e2e/test_catalogue_category.py:155]: I am a little confused as to why this is returning a list of one unit dict as well as a a unit dict?

File: test/e2e/test_catalogue_category_property.py [hunk 29]
```diff
@@ -192,15 +212,16 @@ def post_catalogue_item_property(self, catalogue_item_property_post, catalogue_c
         self._catalogue_item_post_response = self.test_client.post(
             "/v1/catalogue-categories/"
             f"{catalogue_category_id if catalogue_category_id else self.catalogue_category['id']}/properties",
-            json=catalogue_item_property_post,
+            json=replace_unit_values_with_ids_in_properties([catalogue_item_property_post], self.units)[0],
```
Comment (VKTB) [inline test/e2e/test_catalogue_category_property.py:215]: This a bit a difficult to follow so maybe something like this
```python
catalogue_item_property_post = replace_unit_values_with_ids_in_properties([catalogue_item_property_post], self.units)[0]
self._catalogue_item_post_response = self.test_client.post(
    "/v1/catalogue-categories/"
    f"{catalogue_category_id if catalogue_category_id else self.catalogue_category['id']}/properties",
    json=catalogue_item_property_post,
)
```

File: test/e2e/test_catalogue_item.py [hunk 30]
```diff
@@ -109,13 +111,55 @@
...
 
 
+# pylint: disable=duplicate-code
+def _post_units(test_client):
+    """Utility function for posting all mock units defined at the top of this file"""
+
+    response = test_client.get("/v1/units")
+    units = response.json()
+
+    if response.json() == []:
```
Comment (VKTB) [inline test/e2e/test_catalogue_item.py:121]: Why is this needed?

File: test/e2e/test_item.py [hunk 31]
```diff
@@ -82,18 +84,61 @@
...
 
 
+# pylint: disable=duplicate-code
+def _post_units(test_client):
+    """Utility function for posting all mock units defined at the top of this file"""
+
+    response = test_client.get("/v1/units")
+    units = response.json()
+
+    if response.json() == []:
```
Comment (VKTB) [inline test/e2e/test_item.py:94]: Same here

File: test/e2e/test_catalogue_item.py (hunk 30 above)
Comment (joshuadkitenge) [reply test/e2e/test_catalogue_item.py:121]: this is needed if `_post_catalogue_category_with_units` function is ran twice

File: test/e2e/test_catalogue_category.py (hunk 28 above)
Comment (joshuadkitenge) [reply test/e2e/test_catalogue_category.py:155]: I think it makes the tests more readable as I've used unit_dict["mm"] instead of units[0]["id"]

File: test/unit/services/test_catalogue_category_property.py
[the remark this answers is not in the document]
Comment (joshuadkitenge) [reply test/unit/services/test_catalogue_category_property.py:1]: I've already done that test

[the remark this answers is not in the document]
Comment (VKTB) [reply test/unit/services/test_catalogue_category_property.py:1]: Sorry do you mind linking me to it as I couldn't see a unit test in this module that does that?

File: test/e2e/test_catalogue_category.py (hunk 28 above)
Comment (VKTB) [reply test/e2e/test_catalogue_category.py:155]: Maybe for that part but I am not convinced that is the case for the rest, sorry. I don't think the return makes sense and what would happen if in future we want this test method to post more than one unit? Which one would we return as `unit_dict`? This will likely require the return to be changed which will break all the tests that do `unit_dict["mm"]`.

The two ways I can think of that this can be improved is by changing the method to only return `units` and instead of doing `_, unit_dict = _post_units(test_client)` in the tests you would do `unit = _post_units(test_client)[0]`. Alternatively, follow what @joelvdavies did in `_post_catalogue_categories` (return just the unit itself) and instead of `units, _ = _post_units(test_client)` you would do `unit_a = _post_units(test_client)`. You would then put it in a list when passing it to the `replace_unit_values_with_ids_in_properties` method.

Comment (joshuadkitenge) [reply test/e2e/test_catalogue_category.py:155]: It should be `units_dict ` instead of `unit_dict` and if there is more then one then it would just be added into the dictionary `units_dict = {unit_mm["value"]: unit_mm["id"],unit_cm["value"]: unit_cm["id"]}`

**Review by joelvdavies — DISMISSED** [review]
Do not merge until successfully running on dev.

File: test/unit/services/test_catalogue_category_property.py [hunk 32]
```diff
@@ -169,11 +184,55 @@ def test_create_mandatory_property_without_default_value(
...
+    unit_repository_mock,
+    catalogue_category_property_service,
+):
+    """
+    Test creating a property at the catalogue category with a non existent unit id
+    """
+    catalogue_category_id = str(ObjectId())
+    unit = UnitOut(id=str(ObjectId()), **UNIT_A)
+    catalogue_item_property = CatalogueItemPropertyPostRequestSchema(
+        name="Property A", type="number", unit_id=unit.id, mandatory=False
```
Comment (VKTB) [inline test/unit/services/test_catalogue_category_property.py:201]: This can just be `unit_id` as that is the only thing required.
```suggestion
    unit_id = str(ObjectId())
    catalogue_item_property = CatalogueItemPropertyPostRequestSchema(
        name="Property A", type="number", unit_id=unit_id, mandatory=False
```

File: test/unit/services/test_catalogue_category_property.py [hunk 33]
```diff
@@ -169,11 +184,55 @@ def test_create_mandatory_property_without_default_value(
...
+
+    # Mock the stored catalogue category to one without a property with the same name
+    test_helpers.mock_get(catalogue_category_repository_mock, stored_catalogue_category)
+
+    # Mock `get` to return the unit
+    test_helpers.mock_get(unit_repository_mock, None)
+
+    with pytest.raises(MissingRecordError) as exc:
+        catalogue_category_property_service.create(catalogue_category_id, catalogue_item_property)
+    assert str(exc.value) == f"No unit found with ID: {unit.id}"
```
Comment (VKTB) [inline test/unit/services/test_catalogue_category_property.py:222]:
```suggestion
    assert str(exc.value) == f"No unit found with ID: {unit_id}"
```

File: test/unit/services/test_catalogue_category_property.py [hunk 34]
```diff
@@ -169,11 +184,55 @@ def test_create_mandatory_property_without_default_value(
...
+        parent_id=None,
+        catalogue_item_properties=[],
+        created_time=MODEL_MIXINS_FIXED_DATETIME_NOW,
+        modified_time=MODEL_MIXINS_FIXED_DATETIME_NOW,
+    )
+
+    # Mock the stored catalogue category to one without a property with the same name
+    test_helpers.mock_get(catalogue_category_repository_mock, stored_catalogue_category)
+
+    # Mock `get` to return the unit
```
Comment (VKTB) [inline test/unit/services/test_catalogue_category_property.py:217]:
```suggestion
    # Mock `get` to not return a unit
```

**Review by VKTB — APPROVED** [review]
Squash and merge when ready to be merged

**Review by VKTB — APPROVED** [review]

**Review by joelvdavies — APPROVED** [review]
